tests/database/test_header_db.py

```diff
@@ -325,6 +465,210 @@ def test_gap_tracking(headerdb, genesis_header, written_headers, evolving_gaps):
+    )
+))
+@example([(True, 2), (True, 4), (False, 4)])
+def test_gap_continuity(changes):
+    MAX_BLOCKS = 20
+
+    @to_set
+    def _all_missing(chain_gaps):
+        """List all the missing headers, the block numbers in gaps"""
+        gap_ranges, tail = chain_gaps
+        for low, high in gap_ranges:
+            yield from range(low, high + 1)
+
+        yield from range(tail, MAX_BLOCKS)
+
+    def _all_inserted(chain_gaps):
+        """List all the inserted headers, the block numbers not in gaps"""
+        missing = _all_missing(chain_gaps)
+        return set(block_num for block_num in range(MAX_BLOCKS) if block_num not in missing)
+
+    chain_gaps = GENESIS_CHAIN_GAPS
+
+    for do_insert, block_num in changes:
+        starts_inserted = _all_inserted(chain_gaps)
+        starts_missing = _all_missing(chain_gaps)
+
+        if do_insert:
+            to_insert = block_num
+            _, chain_gaps = calculate_gaps(to_insert, chain_gaps)
+            assert not is_block_number_in_gap(to_insert, chain_gaps)
+
+            finished_inserted = _all_inserted(chain_gaps)
+            new_inserts = finished_inserted - starts_inserted
+            if block_num in starts_inserted:
+                assert new_inserts == set()
+            else:
+                assert new_inserts == {block_num}
+
+            finished_missing = _all_missing(chain_gaps)
+            assert to_insert not in finished_missing
+            new_missing = finished_missing - starts_missing
+            assert new_missing == set()
+        else:
+            to_remove = block_num
+            # Note that removing a header is inserting a gap
+            chain_gaps = insert_gap(to_remove, chain_gaps)
+            assert is_block_number_in_gap(to_remove, chain_gaps)
+
+            finished_inserted = _all_inserted(chain_gaps)
+            new_inserts = finished_inserted - starts_inserted
+            assert new_inserts == set()
+
+            finished_missing = _all_missing(chain_gaps)
+            new_missing = finished_missing - starts_missing
+            if block_num in starts_missing:
+                assert new_missing == set()
+            else:
+                assert new_missing == {block_num}
+
+        # Validate the gap invariants:
+
+        # 1. gaps are sorted
+        for low, high in chain_gaps[0]:
+            assert high >= low, chain_gaps
+
+        # 2. gaps are not contiguous
+        for low_range, high_range in sliding_window(2, chain_gaps[0]):
+            # the top of the low range must not be sequential with the bottom of the high range
+            assert low_range[1] + 1 < high_range[0], chain_gaps
+
+        # 3. final gap is not contiguous with the tail
+        if len(chain_gaps[0]):
+            final_gap_range = chain_gaps[0][-1]
+            assert final_gap_range[1] + 1 < chain_gaps[1], chain_gaps
+
+
+@given(st.data())
+@settings(max_examples=200)
+def test_true_chain_canonicalized_regardless_of_order(genesis_header, data):
```

Comment:
This was the test that drove the whole PR forward.

The core idea is that we should be able to checkpoint A at any location, and insert headers for A in any valid order (from genesis or checkpoint forward), and insert B in any valid order. Then after all chain A headers are inserted once, then the whole A chain is canonical.

I suppose that belongs in a docstring ^.